Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
-	float invMassB;
-	float normMassA;
-	float normMassB;
-
-    hitMOID = g_NoMOID;
-    currentMOID = g_NoMOID;
-    pIntersectedMO = 0;
-    exitDirection.Reset();
-    atomExitVector .Reset();
-    totalExitVector.Reset();
-    intersectingAtoms.clear();
-
-    // First go through all atoms to find the first intersection and get the intersected MO
-    for (Atom *atom : m_Atoms)
-    {
-        atomOffset = atom->GetOffset().GetXFlipped(m_pOwnerMO->IsHFlipped());
-        atomOffset *= rotation;
-        atom->SetupPos(position + atomOffset);
-
-        atomPos = atom->GetCurrentPos();
-		hitMOID = g_SceneMan.GetMOIDPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
-		// Don't count MOIDs ignored
-		if (hitMOID != g_NoMOID && !atom->IsIgnoringMOID(hitMOID))
-		{
-			// Save the correct MOID to search for other atom intersections with
-			currentMOID = hitMOID;
-
-			// Get the MO we seem to be intersecting
-			MovableObject *tempMO = g_MovableMan.GetMOFromID(hitMOID);
-			RTEAssert(tempMO, "Intersected MOID couldn't be translated to a real MO!");
-			tempMO = tempMO->GetRootParent();
-
-			if (tempMO->GetsHitByMOs())
-			{
-				// Make that MO draw itself again in the MOID layer so we can find its true edges
-				pIntersectedMO = tempMO;
-				pIntersectedMO->Draw(g_SceneMan.GetMOIDBitmap(), Vector(), g_DrawMOID, true);
-				break;
-				// TODO: Remove
-				//                g_FrameMan.SaveBitmapToBMP(g_SceneMan.GetMOIDBitmap(), "MOIDMap");
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool AtomGroup::ResolveTerrainIntersection(Vector &position, unsigned char strongerThan) const {
+		std::list<Atom *> intersectingAtoms;
+		MOID hitMaterial = g_MaterialAir;
+
+		float strengthThreshold = (strongerThan != g_MaterialAir) ? g_SceneMan.GetMaterialFromID(strongerThan)->GetIntegrity() : 0.0F;
+		bool rayHit = false;
+
+		Vector atomOffset = Vector();
+		Vector atomPos = Vector();
+
+		// First go through all Atoms to find the first intersection and get the intersected MO
+		for (Atom *atom : m_Atoms) {
+			atomOffset = m_OwnerMO->RotateOffset(atom->GetOffset());
+			atom->SetupPos(position + atomOffset);
+			atomPos = atom->GetCurrentPos();
+			hitMaterial = g_SceneMan.GetTerrain()->GetPixel(atomPos.GetFloorIntX(), atomPos.GetFloorIntY());
+			if (hitMaterial != g_MaterialAir && strengthThreshold > 0.0F && g_SceneMan.GetMaterialFromID(hitMaterial)->GetIntegrity() > strengthThreshold) {
+				intersectingAtoms.push_back(atom);
 			}
 		}
-    }
+		if (intersectingAtoms.empty()) {
+			return true;
+		}
 
-    // If no intersected MO is found, we are clear and can quit
-	if (!pIntersectedMO) {
-		return false;
-	}
+		// If all Atoms are intersecting, we're screwed?!
+		if (intersectingAtoms.size() >= m_Atoms.size()) {
+			return false;
+		}
+
+		Vector exitDirection = Vector();
+
+		// Go through all intersecting Atoms and find their average inverse normal
```

Comment:
Is it their inverse normal or just their normal? I'm on the fence about this comment, I don't think it's actually useful as-is, but a comment/comments actually explaining in english what all this stuff is doing would probably be nice. Might be tough to figure out and write though, maybe a TODO to put in such a comment would be fine.

Reply:
This comment applies to the next ~10 lines where `exitDirection = -exitDirection;` so I guess it *is* inverse.

Reply:
Haha okay, maybe that's what it means. We can leave it be for now and get rid of it when the method is broken into pieces and stuff.